### DISCUSSION TITLE
Ellipse in Pagination for Dynamic data

### DISCUSSION THREAD

**Discussion by shivamBentley** [iTwin/iTwinUI#2081] [Q&A]
Hi,

I am using paginator in ITwin UI Table, I am facing an issue to show ellipses in paginator for dynamic data. Assume we are fetching data from backed only when it is required. so we made the logic if user click on last page ( in paginator ) we start fetching data and update in table it, we try to load one page extra. [Note: We know we have data or not on before every fetch].

if more data is available
![image](https://github.com/iTwin/iTwinUI/assets/110664877/39ae5bb4-af24-4708-9db9-c920c29d699e)

if more data is not available
![image](https://github.com/iTwin/iTwinUI/assets/110664877/2d3cebde-ce0d-4b09-8864-436c9eeb6047)

We want to make users experience bit better, If we have data can we show ellipses after last page if we have data if not then show last page number ( as default behaviour )

**Comment by r100-stack** [comment]
Our current default logic to show the ellipses between the current pages range and the first or last page. We do not show ellipses before the first page or after the last page.

E.g. [Table → With Paginator](https://itwin.github.io/iTwinUI/react/?story=table--with-paginator&theme=dark) story:

<img width="774" alt="Table → With Paginator story" src="https://github.com/iTwin/iTwinUI/assets/45748283/0220ed3a-9c2b-48ad-a2ef-420242e95858">

Here, we show ellipses between pages 1 and 13 and also between 27 and 101.

---

To make the `TablePaginator` work properly by default even when lazy-loading, we need to pass the `totalRowCount` prop to the `TablePaginator`. Once the page is clicked (TablePaginator's `onPageChange`), you can lazily download the page data. You can take ideas from the [Table → With manual paginator and filter story](https://itwin.github.io/iTwinUI/react/?story=table--with-manual-paginator-and-filter&theme=dark) for implementing this.

**Reply by shivamBentley** [reply]
I understood your point, but we don't know `totalRowCount` which makes it a bit difficult to control the paginator.

**Reply by mayank99** [reply]
Can you share more details about your user case? `totalRowCount` sounds like something that should be possible to query, no? Also, does this mean the user will never reach the last page? And how does sorting work?

**Reply by shivamBentley** [reply]
Please check out this code - [Table - With manual paginator](https://itwin.github.io/iTwinUI/react/?source=true&story=table--with-manual-paginator-and-filter&theme=dark#ladle_loc_1908)


At line number `2080`, here we are passing `total size` of `data`. so, we do not have this information we are implementing `server-side` pagination in our list. 
![image](https://github.com/iTwin/iTwinUI/assets/110664877/05882d86-3226-46c7-9b2b-152f2038e79f)

When user changes pageSize, page, sort or search we query data using these informations. So, we get only two information in response `data` and the `next cursor`. if `next cursor` is `null`, it means we don't have any `data`, else we have `data`. If we have more data, then we want to show `paginator` like this.
![image](https://github.com/iTwin/iTwinUI/assets/110664877/cbfc8293-aea8-4ff3-98c0-ad107b67d422)


@mayank99

**Reply by r100-stack** [reply]
I'm still a bit confused why the `totalRowsCount` would not be known on the frontend. Could you elaborate?

Based on the table filters, would it not be possible to know the number of matching rows on the backend? This backend query could be a separate one that is not related to the `data` and `next cursor` call that you're making. i.e. you'd still be lazily downloading the page data once a page button clicked. But you'll be making a separate query in the beginning to just get the number of matching rows.

> When user changes pageSize, page, sort or search we query data using these informations. So, we get only two information in response `data` and the `next cursor`.

Changing "pageSize, page, sort" wouldn't change the total row count. But actions like search, filter, etc. could change the total row count. So, whenever an action is performed that could change the total row count, would it be possible to send a new query to the backend to get the number of rows that match the new filters, search query, etc.?

Additionally, showing the user the total number of pages would also be better UX. They can then quickly navigate to the first/last page.
